src/OrchardCore/OrchardCore.Modules/ModularTenantRouterMiddleware.cs

```diff
@@ -49,37 +51,33 @@ public async Task Invoke(HttpContext httpContext)
                 httpContext.Request.Path = httpContext.Request.Path.ToString().Substring(httpContext.Request.PathBase.Value.Length);
             }
 
-            // TODO: Invalidate the pipeline automatically when the shell context is changed
-            // such that we can reload the middlewares and the routes. Implement something similar
-            // to IRunningShellTable but for the pipelines.
-
             // Do we need to rebuild the pipeline ?
-            var rebuildPipeline = httpContext.Items["BuildPipeline"] != null;
-            if (rebuildPipeline && _pipelines.ContainsKey(shellContext.Settings.Name))
+            if (!shellContext.IsPipelineBuilt)
             {
-                _pipelines.Remove(shellContext.Settings.Name);
-            }
+                var semaphore = _semaphores.GetOrAdd(shellContext.Settings.Name, (name) => new SemaphoreSlim(1));
 
-            RequestDelegate pipeline;
+                // Building a pipeline for a given shell can't be done by two requests.
+                await semaphore.WaitAsync();
 
-            if (!_pipelines.TryGetValue(shellContext.Settings.Name, out pipeline))
-            {
-                // Building a pipeline can't be done by two requests
-                lock (_pipelines)
+                try
                 {
-                    if (!_pipelines.TryGetValue(shellContext.Settings.Name, out pipeline))
+                    if (!shellContext.IsPipelineBuilt)
                     {
-                        pipeline = BuildTenantPipeline(shellContext.ServiceProvider, httpContext.RequestServices);
+                        _pipelines[shellContext.Settings.Name] = BuildTenantPipeline(
```

Comment:
You don't need a CD for that, just declare a local variable after the `if (!shellContext.IsPipelineBuilt)` and use it on line 80

Reply:
The CD is to cache pipelines per shell, and we use it line 80 for a given shell even line 66 was not executed because its pipeline is already built.

Also, line 66 can be executed only once at a time for a given shell, but can be executed concurrently by multiple shells.